### PR TITLE
Getting data from google via backend

### PR DESCRIPTION
**Pull request by Panton1986**
Added file "google_auth_controller"
Added route: {"rule": "/checking", "methods": ["POST"], "view_func": google_auth_controller.checking},

### CODE REVIEW

File: app/controllers/google_auth_controller.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    token_request = google.auth.transport.requests.Request(
+        session=cached_session)
+
+    id_info = id_token.verify_oauth2_token(
+        request=token_request,
+        audience=google_client_id,
+        id_token=credentials._id_token
+    )
+
+    google_id = id_info.get("sub")
```
Comment (denis-progman) [inline app/controllers/google_auth_controller.py:46]: you don't need these vars, you can put it directly into the 'response' object below

File: app/controllers/user_controller.py [hunk 2]
```diff
@@ -36,4 +45,57 @@ def update_user(user_id):
...
             UserService.delete_user(user_id=user_id)
-        )
\ No newline at end of file
+        )
+    
+    def checking():
+        # Get the string from the frontend
+        data = request.get_json()
+        input_string = data.get('code')
+        auth_code = data.get('serverAuthCode')
```
Comment (denis-progman) [inline app/controllers/user_controller.py:54]: better to make 'serverAuthCode' and 'code' and 'input_string' and 'SECRET_CODE' class's constants

**Review by denis-progman — CHANGES_REQUESTED** [review]
Please fix and comment back all the  comments. We can discuss the details in the dev chat.

File: app/controllers/google_auth_controller.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+from flask import abort, session, request, redirect
+from configs import Config
+
+from google.oauth2 import id_token
+from google_auth_oauthlib.flow import Flow
+from pip._vendor import cachecontrol
+import google.auth.transport.requests
+
+from flask import request
+
```
[the remark this answers is not in the document]
Comment (Panton1986) [reply app/controllers/google_auth_controller.py:14]: Done

File: app/controllers/google_auth_controller.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    pathlib.Path(__file__).parent, "client_secret.json")
+
+flow = Flow.from_client_secrets_file(
+    client_secrets_file=client_secrets_file,
+    scopes=["https://www.googleapis.com/auth/userinfo.profile",
+            "https://www.googleapis.com/auth/userinfo.email", "openid"],
+    redirect_uri="http://127.0.0.1:8888/callback"
+)
+
+def checking():
```
[the remark this answers is not in the document]
Comment (Panton1986) [reply app/controllers/google_auth_controller.py:28]: Done

File: app/controllers/user_controller.py [hunk 5]
```diff
@@ -36,4 +45,57 @@ def update_user(user_id):
     def delete_user(user_id):
         return jsonify(
             UserService.delete_user(user_id=user_id)
-        )
\ No newline at end of file
+        )
+    
+    def checking():
```
[the remark this answers is not in the document]
Comment (Panton1986) [reply app/controllers/user_controller.py:50]: Done

File: app/controllers/user_controller.py [hunk 6]
```diff
@@ -36,4 +45,57 @@ def update_user(user_id):
...
+        userid = credentials.id_token['sub']
+        email = credentials.id_token['email']
+        
+        response = {
+            'hashed_code': hashed_code,
+            'userid': userid,
+            'email': email
+        }
+        # Hash the special code using SHA-256
+        session['key'] = hashlib.sha256(special_code.encode()).hexdigest()
```
[the remark this answers is not in the document]
Comment (Panton1986) [reply app/controllers/user_controller.py:100]: Done

File: app/controllers/user_controller.py (hunk 2 above)
Comment (Panton1986) [reply app/controllers/user_controller.py:54]: Done

File: app/controllers/user_controller.py [hunk 7]
```diff
@@ -36,4 +45,57 @@ def update_user(user_id):
...
+        }
+
+        # If this request does not have `X-Requested-With` header, this could be a CSRF
+
+        # if not request.headers.get('X-Requested-With'):
+        #     abort(403)
+
+        # Set path to the Web application client_secret_*.json file you downloaded from the
+        # Google API Console: https://console.cloud.google.com/apis/credentials
+        CLIENT_SECRET_FILE = os.path.join(pathlib.Path(__file__).parent, "client_secret.json")
```
[the remark this answers is not in the document]
Comment (Panton1986) [reply app/controllers/user_controller.py:78]: Done

File: app/controllers/google_auth_controller.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+from pip._vendor import cachecontrol
+import google.auth.transport.requests
+
+from flask import request
+
+os.environ["OAUTHLIB_INSECURE_TRANSPORT"] = "1"
+
+google_client_id = Config.GOOGLE_CLIENT_ID
+client_secrets_file = os.path.join(
+    pathlib.Path(__file__).parent, "client_secret.json")
```
[the remark this answers is not in the document]
Comment (Panton1986) [reply app/controllers/google_auth_controller.py:19]: Done

File: app/controllers/google_auth_controller.py [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+from configs import Config
+
+from google.oauth2 import id_token
+from google_auth_oauthlib.flow import Flow
+from pip._vendor import cachecontrol
+import google.auth.transport.requests
+
+from flask import request
+
+os.environ["OAUTHLIB_INSECURE_TRANSPORT"] = "1"
```
[the remark this answers is not in the document]
Comment (Panton1986) [reply app/controllers/google_auth_controller.py:15]: moved to the config

File: app/controllers/google_auth_controller.py (hunk 1 above)
Comment (Panton1986) [reply app/controllers/google_auth_controller.py:46]: object creates in return directly

File: app/.venv/bin/Activate.ps1
Comment (denis-progman) [inline app/.venv/bin/Activate.ps1:1]: please remove the .venv folder, this is trash. Add it to git ignore when run git rm -r app/.venv and commit it (removes the folder from your computer as well, but I believe you don't need it, we use docker...)
then make sure that .venv is disappeared and do 'push'.

File: app/router.py [hunk 10]
```diff
@@ -3,10 +3,14 @@
 from controllers import GamerController
```
[the remark this answers is not in the document]
Comment (denis-progman) [reply app/router.py:3]: ???

**Review by denis-progman — COMMENTED** [review]
Plese comment all the review comments

File: app/controllers/google_auth_controller.py [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+
+google_client_id = Config.GOOGLE_CLIENT_ID
+client_secrets_file = os.path.join(
+    pathlib.Path(__file__).parent, "client_secret.json")
+
+flow = Flow.from_client_secrets_file(
+    client_secrets_file=client_secrets_file,
+    scopes=["https://www.googleapis.com/auth/userinfo.profile",
```
[the remark this answers is not in the document]
Comment (denis-progman) [reply app/controllers/google_auth_controller.py:23]: ???

File: app/controllers/google_auth_controller.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (denis-progman) [reply app/controllers/google_auth_controller.py:28]: Are you sure you have created class Service for Google auth?  and there is no class ion the controller part as I see.. Check it out please.

File: app/controllers/user_controller.py [hunk 12]
```diff
@@ -36,4 +45,57 @@ def update_user(user_id):
...
+
+        # Exchange auth code for access token, refresh token, and ID token
+        credentials = client.credentials_from_clientsecrets_and_code(
+            CLIENT_SECRET_FILE,
+            ['https://www.googleapis.com/auth/drive.appdata', 'profile', 'email'],
```
[the remark this answers is not in the document]
Comment (denis-progman) [reply app/controllers/user_controller.py:83]: ???

**Review by denis-progman — COMMENTED** [review]
don't forget removing .venv

File: app/controllers/google_auth_controller.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Panton1986) [reply app/controllers/google_auth_controller.py:28]: I've checked it out. Classes have created for every service and controller

File: app/router.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Panton1986) [reply app/router.py:3]: All unused file deleted

File: app/.venv/bin/Activate.ps1
Comment (Panton1986) [reply app/.venv/bin/Activate.ps1:1]: Done

File: app/controllers/google_auth_controller.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Panton1986) [reply app/controllers/google_auth_controller.py:23]: Gentle information moved to .env and config files. Strings moved to class' constants

**Review by denis-progman — APPROVED** [review]

**Comment by denis-progman** [comment]
cool, thanks!

**Pull request merged**